File: leetcodes/300. Longest Increasing Subsequence/lengthOfLIS.java
class Solution {
  public int binarySearchPosition(int[] sub, int val, int len) {
    int low = 0;
    int high = len - 1;
    while (low <= high) {
      int mid = low + (high - low) / 2;
      if (sub[mid] < val)
        low = mid + 1;
      else if (val < sub[mid])
        high = mid - 1;
      else
        return mid;
    }
    return low;
  }

  public int lengthOfLIS(int[] nums) {
    // dp -> use sub[] to get the correct length (but not correct elements)
    // O(n*m), m is len(sub) -> use while (pos <= sub_len) to get the pos and the
    // sub array
    // O(nlogn), use binary search to find the correct insertion position,
    // space O(n)
    int n = nums.length;
    int len = 0;
    int[] sub = new int[n];
    sub[len++] = nums[0];
    for (int i = 0; i < n; i++) {
      if (nums[i] > sub[len - 1]) {
        sub[len++] = nums[i];
      } else {
        int pos = binarySearchPosition(sub, nums[i], len);
        sub[pos] = nums[i];
      }
    }
    return len;
  }
}

File: leetcodes/300. Longest Increasing Subsequence/lengthOfLIS.js
/**
 * @param {number[]} nums
 * @return {number}
 */
 var lengthOfLIS = function(nums) {
  //time O(n^2), space O(n)
  let DP = new Array(nums.length).fill(1)
  let result = 1
  for (let i = 1; i < nums.length; i ++) {
    for (let j = 0; j < i; j ++) {
      if (DP[i] > DP[j]) {
        DP[i] = Math.max(DP[j] + 1, DP[i])
      }
    }

  }
  return Math.max.apply(Math, DP) //get the max from the array
};
var lengthOfLIS = function(nums) {
  //time O(nlogn), space O(n) -> binary search for replacing sub element
  const BSP = (sub, val) => {
    let low = 0
    let high = sub.length - 1
    while (low <= high) {
      let mid = low + Math.floor((high - low)/2)
      if (sub[mid] < val) {
        low = mid + 1
      } else if (sub[mid] > val) {
        high = mid - 1
      } else return mid
     }
      return low
    }
  let sub = [];
  for (let val of nums) {
    let pos = BSP(sub, val)
    if (pos === sub.length) sub.push(val)
    else sub[pos] = val
  }
  //[2,3,7,18] --> sub
  return sub.length
  }